todo-list-backend/src/index.ts

```diff
@@ -0,0 +1,124 @@
+import {decode, sign} from "jsonwebtoken";
+import {applyMiddleware} from "graphql-middleware";
+import {ApolloServer, gql} from "apollo-server";
+import {makeExecutableSchema} from "graphql-tools";
+
+import {User} from "./types";
+import {InMemoryStorage, IStorage} from "./data"
+import {buildPermissions} from "./permissions";
+
+const typeDefs = gql`
+    type User {
+        name: String
+    }
+
+    type ToDo {
+        id: Int
+        name: String
+        done: Boolean
+        assignee: User
+    }
+
+    type Query {
+        readTodos: [ToDo]!
+    }
+
+    type Mutation {
+        login(name: String!, password: String!): String
+        createTodo(name: String!): ToDo
+        updateTodoName(id: Int!, name: String!): ToDo
+        updateTodoDone(id: Int!, done: Boolean!): ToDo
+        deleteTodo(id: Int!): ToDo
+        createUser(name: String! password: String!): User
+    }
+`;
+
+function buildResolvers(storage: IStorage, authSecret: string): any {
+    return {
+        Query: {
+            readTodos: (parent: any, args: any, context: any, info: any) => {
+                let user: User = context.user;
+                return storage.readTodos(user);
+            },
+        },
+
+        Mutation: {
+            login: (parent: any, args: any, context: any, info: any) => {
+                let user = storage.readUser(args.name);
+
+                if (!user || !user.checkPassword(args.password)) {
+                    return null;
+                }
+
+                let payload = {
+                    username: user.name
+                };
+
+                return sign(payload, authSecret, {expiresIn: "1 day"});
+            },
+            createTodo: (parent: any, args: any, context: any, info: any) => {
+                let user: User = context.user;
+                return storage.createTodo(user, args.name);
```

Comment:
Instead of putting the storage into the initializer function, you could hand it over to `context. Your resolver would look like:
```suggestion
                const { storage } = context
                return storage.createTodo(user, args.name);
```